_studio/mfx_lib/decode/vp9/src/mfx_vp9_dec_decode_hw.cpp

```diff
@@ -85,6 +85,87 @@ bool CheckHardwareSupport(VideoCORE *p_core, mfxVideoParam *p_video_param)
     return true;
 }
 
+class ReferenceFrameStorage
+{
+private:
+    std::vector<UMC_VP9_DECODER::VP9DecoderFrame> m_submittedFrames;
+    mfx_UMC_FrameAllocator *m_frameAllocator;
+
+    void FreeReferenceFrames(UMC::FrameMemID frameId)
+    {
+        auto find_it = std::find_if(m_submittedFrames.begin(), m_submittedFrames.end(), [frameId](UMC_VP9_DECODER::VP9DecoderFrame frame){ return (frame.currFrame == frameId); });
+        if (find_it != m_submittedFrames.end())
+        {
+            for (mfxI32 ref_index = 0; ref_index < NUM_REF_FRAMES; ++ref_index)
+            {
+                const UMC::FrameMemID mid = find_it->ref_frame_map[ref_index];
+                if (mid >= 0 && mid != find_it->currFrame)
+                {
+                    m_frameAllocator->DecreaseReference(mid);
```

Comment:
IncreaseReference outside Storage (Decoder UpdateReference) DecreaseReference inside Sotrage class.
I think it would be better to put reference counting logic in one place this Storage class, add new frame increases all required references, mark frame as decoded, free unused frames decreases all reference back and it should be maid from synchronous code DecodeFrameAsync, move all DecreaseReference functions from async DecodeRoutine to the decoder function.